controllers: support addon artifact images

ANJANA-ARK opened the pull request:
Add support for installing custom kernel/initrd artifacts from container images in DaemonSet mode. Introduces osc-kata-addons-install.sh script and ConfigMap-based configuration (kata-addon-artifacts).

Reuses extract_container_image from lib.sh. Minimal changes to existing flows. Provider-specific config updates (configuration-se.toml, configuration-tdx.toml, etc.) based on ConfigMap provider field.

beraldoleal commented on config/baremetal/kata-addon-artifacts.yaml:
```diff
@@ -0,0 +1,20 @@
+  name: kata-addon-artifacts
+  namespace: openshift-sandboxed-containers-operator
+data:
+  addonImage: "quay.io/<username>/kata-se-artifacts:v1.0"
```
username?

ANJANA-ARK replied:
I've changed it to quay.io/openshift_sandboxed_containers/kata-se-artifacts:v1.0 as sample image

beraldoleal commented on config/baremetal/kata-addon-artifacts.yaml:
```diff
@@ -0,0 +1,20 @@
+apiVersion: v1
+kind: ConfigMap
```
are we introducing this new cm or should we use kataconfig? @gkurz any comments here?

gkurz replied:
Good question. CM is definitely more flexible and easier to put in place but offers no validation, nor versioning. So I would say it depends on how far you look. If the main goal is to release fast at the expense of stability and maintainability, then go CM. If the goal is to have rock solid data structures that provide consistency and stability for the long term, then go CR.

At this point, it seems that the project is clearly driven by short term so I'd stick to CM.

beraldoleal replied:
Hi @gkurz , thanks for chiming in..

I get were you came from... but I’m still unsure why this needs a separate ConfigMap instead of extending KataConfig.

We can release fast with a ConfigMap, but that creates a second configuration surface with no validation, no upgrade guarantees, and no clear lifecycle. KataConfig already defines the operator’s configuration API, and adding new fields there isn´t additive? iiuc, it doesn’t require a new CRD version, right?

One more point: is this DP, TP, or GA code? If it’s not GA, I’m fine keeping it as a ConfigMap for now (Running the risk of us keeping this forever: ( ). But if it’s GA, we should think carefully before asking users to edit config in two different places and losing the CRD benefits. This tends to come back to us sooner or later. We can still move fast if we do things in the right place.

gkurz replied:
> iiuc, it doesn’t require a new CRD version, right?

Not familiar enough with CRDs to know when versioning is mandatory.

> One more point: is this DP, TP, or GA code? If it’s not GA, I’m fine keeping it as a ConfigMap for now (Running the risk of us keeping this forever: ( ). 

This is for bm coco on IBM-z, thus TP.

beraldoleal replied:
Ok, thanks @gkurz . 

@ANJANA-ARK we can close this thread, but I would suggest you to consider keeping that in mind for for the GA.

ANJANA-ARK replied:
sure

beraldoleal commented on config/baremetal/kata-addon-artifacts.yaml:
```diff
@@ -0,0 +1,20 @@
+  initrdPath: "/artifacts/initrd/kata-containers-initrd.img"
+  
+  # Define the OSC version
+  version: "v1.11"
```
Do you mind expanding a bit on the comments why this version is needed? In which situation we could have OSC version installed X and here using Y?

ANJANA-ARK replied:
We added versions so that we'll use it in upgrade function to upgrade OSC, but now as discussed above https://github.com/openshift/sandboxed-containers-operator/pull/1334#discussion_r2527488195 I've removed upgrade option, so I've removed this param

Pacho20 commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -211,18 +210,38 @@ main() {
+
+		sleep infinity
 		;;
+	upgrade)
```
Could you explain how this would be executed in the script? Currently, the controller cannot set upgrade as the action for the DaemonSet. Do you have plans to extend the controller to handle that?

ANJANA-ARK replied:
Hi @Pacho20 
We planned to add an upgrade option so the controller could compare the running OSC version and perform a version upgrade. Since the DaemonSet controller currently does not support upgrade actions, for now I’ll remove the upgrade option to avoid confusion.
Do you happen to know when support for upgrade actions might be added to the controller?

Pacho20 replied:
I don’t have any information about adding an upgrade action to the controller. Nevertheless, the installation action is designed to handle upgrades by checking whether the latest versions of the packages are installed. If they aren’t, it reinstalls them with the newer versions.

beraldoleal commented on controllers/daemonset_reconcile.go:
```diff
@@ -672,6 +681,97 @@ func (r *KataConfigOpenShiftReconciler) daemonSetForKataInstall(action KataDaemo
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// ConfigMap not found - addon not configured
```
Do plan to add a specific log msg here? If not, this if is unnecessary.

Pacho20 commented on scripts/kata-install/osc-kata-addons-install.sh:
```diff
@@ -0,0 +1,227 @@
+#   0 on success
+#######################################
+upgrade_addons() {
+    local version_file="/usr/share/kata-containers/.addon-version"
```
Both the installation and uninstallation use `"/etc/kata-containers/.addon-version"`. I don’t see `/usr` used anywhere else. I’m not very familiar with Kata - does it create this file? Otherwise, the upgrade would always call `install_addons`.

ANJANA-ARK replied:
As we remove the upgrade option, I've removed this as well

beraldoleal commented on controllers/daemonset_reconcile.go:
```diff
@@ -672,6 +681,97 @@ func (r *KataConfigOpenShiftReconciler) daemonSetForKataInstall(action KataDaemo
+	})
+
+	return envVars
+}
```
I haven't tested my proposal, but maybe a less verbose method, with two structures for optional and mandatory?

```
	var cm corev1.ConfigMap
	if err := r.Client.Get(context.Background(),
		types.NamespacedName{Name: "kata-addon-artifacts", Namespace: OperatorNamespace},
		&cm,
	); err != nil {
		if !errors.IsNotFound(err) {
			r.Log.Error(err, "Failed to get addon ConfigMap")
		}
		return nil
	}

	data := cm.Data

	// --- Mandatory variables ---
	mandatory := []struct {
		key, env string
	}{
		{"addonImage", "ADDON_IMAGE"},
	}

	var envs []corev1.EnvVar
	for _, v := range mandatory {
		val := data[v.key]
		if val == "" {
			return nil // mandatory missing → addon not configured
		}
		envs = append(envs, corev1.EnvVar{Name: v.env, Value: val})
	}

	// --- Optional variables ---
	optional := []struct {
		key, env string
	}{
		{"provider", "ADDON_PROVIDER"},
		{"kernelPath", "ADDON_KERNEL_PATH"},
		{"initrdPath", "ADDON_INITRD_PATH"},
		{"version", "ADDON_VERSION"},
	}

	for _, v := range optional {
		val := data[v.key]
		if v.key == "version" && val == "" {
			val = extractVersionFromImage(data["addonImage"])
		}
		if val != "" {
			envs = append(envs, corev1.EnvVar{Name: v.env, Value: val})
		}
	}

	r.Log.Info("Addon artifacts configured", "image", data["addonImage"])
	return envs
```

gkurz commented on config/baremetal/kata-addon-artifacts.yaml:
```diff
@@ -0,0 +1,20 @@
+  
+  # Required: Provider type (se, tdx, snp, etc.)
+  # This determines which config file to update: kata-<provider>/configuration.toml
+  provider: "se"
```
AFAIK this is for IBM SE only and it is unlikely to change anytime soon. Let's keep it simple and drop that.

ANJANA-ARK replied:
Hi @gkurz , I've removed this param

ajaypvictor replied:
Hi @gkurz @bpradipt @beraldoleal,
As per the initial design plan for this feature, idea was to use this for other providers as well maybe for dev-test, by hardcoding this variable, we are restricting this feature only for IBM SE, as this suggestion requires hardcoding of the SE specific config toml file on the code itself.
Could you please confirm if this approach is fine?
Thanks..

gkurz replied:
Other TEEs are fully handled by the kata-containers RPM. This PR only exists because IBM SE artifacts cannot be shipped the same way. Hardcoding is the way to go.

gkurz commented on scripts/kata-install/osc-kata-addons-install.sh:
```diff
@@ -0,0 +1,227 @@
+        fi
+        echo "  Updated initrd: $initrd_path"
+    fi
+
```
It seems that all this logic could be replaced with a simple config drop in. Something like :

```
cat<<EOF>/etc/kata-containers/kata-se/config.d/coco-guest.toml
[hypervisor.qemu]
kernel = "${kernel_path}"
initrd = "${initrd_path}"
EOF
```

ANJANA-ARK replied:
As per my understanding, https://github.com/kata-containers/kata-containers/blob/main/src/runtime/config/configuration-qemu-se.toml.in this will be present in /etc/kata-containers/kata-se as I'm changed just the kernel/initrd in that file

gkurz replied:
I was suggesting to leverage https://github.com/kata-containers/kata-containers/blob/main/src/runtime/README.md#drop-in-configuration-file-fragments instead of modifying the existing configuration file.  This can be improved later.

ANJANA-ARK replied:
sure I'll make this changes in another PR

beraldoleal commented on scripts/kata-install/Dockerfile:
```diff
@@ -1,15 +1,14 @@
 RUN mkdir -p /scripts
 
-ADD osc-kata-install.sh osc-configs-script.sh osc-log-level.sh lib.sh /scripts/
+ADD osc-kata-install.sh osc-configs-script.sh osc-log-level.sh lib.sh osc-kata-install.sh osc-kata-addons-install.sh /scripts/
```
osc-kata-install.sh is duplicated here.

ANJANA-ARK replied:
Thanks for pointing out I've changed it 👍

beraldoleal reviewed (commented):
Hi @ANJANA-ARK , thanks for the changes, I closed all my previous threads, but I left one minor comment for you. Besides that and the others comments, lgtm.

ANJANA-ARK commented:
Hi @gkurz @beraldoleal @Pacho20 I've addressed the review comments, kindly review. Thank  you!

ajaypvictor reviewed (approved):
LGTM, Thanks Anjana..

gkurz commented on scripts/kata-install/osc-kata-addons-install.sh:
```diff
@@ -0,0 +1,139 @@
+# Update provider-specific configuration file
+# Arguments:
+#   $1: kernel path (empty if not installed)
+#   $2: initrd path (empty if not installed)
```
It looks like a leftover.

ANJANA-ARK replied:
Thanks for pointing out, I've removed this

gkurz commented on scripts/kata-install/osc-kata-addons-install.sh:
```diff
@@ -0,0 +1,139 @@
+    chroot /host cp "$config_file" "${config_file}.backup-$(date +%s)"
+    
+    # Update kernel if provided
+    if [ -n "$kernel_path" ]; then
```
Shouldn't `kernel_path` be mandatory ?

ANJANA-ARK replied:
Yes, I've changed this 👍

gkurz replied:
You dropped the check in the caller but this still looks optional to me... I'll reword my question : shouldn't the script just fail if no kernel is provided ?

ANJANA-ARK replied:
sure I'll add a check

gkurz replied:
Since this is checked in the caller, the `if` isn't needed anymore.

beraldoleal reviewed (approved):
Except from @gkurz comments, lgtm. thanks.

ANJANA-ARK commented:
Hi @gkurz @Pacho20 , I've addressed the review comments can you please review, thank you!

gkurz reviewed (commented):
This part of the commit log is outdated. Please drop it.

```
Provider-specific config updates (configuration-se.toml, configuration-tdx.toml, etc.) based on ConfigMap provider field.
```

gkurz commented on scripts/kata-install/Dockerfile:
```diff
@@ -1,15 +1,14 @@
-FROM registry.access.redhat.com/ubi9/skopeo:9.6-1760517870
+FROM registry.access.redhat.com/ubi9/skopeo:9.6-1756959203
```
This is wrong on several aspects :
- you should never have to downgrade a base image like that as it can degrade the health index of the resulting image
- unless you really need a new base image in your PR, avoid adding an unrelated change like this

According to https://catalog.redhat.com/en/software/containers/ubi9/skopeo/61959869ce09ff98ccc36338?image=68b91a5b1c978d120a3f0f1d and https://catalog.redhat.com/en/software/containers/ubi9/skopeo/61959869ce09ff98ccc36338?image=68b91a5b1c978d120a3f0f1d, both images have a B grade... we should upgrade to something newer, but this will happen in a separate PR.

In the meatime, please drop this change.

ANJANA-ARK replied:
changed this

gkurz commented on controllers/daemonset_reconcile.go:
```diff
@@ -672,6 +682,73 @@ func (r *KataConfigOpenShiftReconciler) daemonSetForKataInstall(action KataDaemo
+	optional := []struct {
+		key, env string
+	}{
+		{"kernelPath", "ADDON_KERNEL_PATH"},
```
Why is the kernel path optional ? Isn't providing a kernel the justification for this PR ?

ANJANA-ARK replied:
sure I've changed this

gkurz commented on scripts/kata-install/osc-kata-addons-install.sh:
```diff
@@ -0,0 +1,140 @@
+source "${SCRIPT_DIR}/lib.sh"
+
+#######################################
+# Update provider-specific configuration file
```
Please drop the `provider` wording everywhere. This is only for IBM SE.

ANJANA-ARK replied:
sure

gkurz commented on scripts/kata-install/osc-kata-addons-install.sh:
```diff
@@ -0,0 +1,140 @@
+    
+    echo "Installing addon artifacts from: $addon_image"
+    
+    local kernel_src="${ADDON_KERNEL_PATH:-}"
```
It looks like a good place to bail out if no kernel path is provided.

gkurz replied:
Same here.

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -109,42 +109,41 @@ install() {
 		fi
 	done
 
-	# If nothing to install, exit early
```
why dropping the comment ?

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -109,42 +109,41 @@ install() {
-	# If nothing to install, exit early
 	if [[ ${#install_rpms[@]} -eq 0 ]]; then
-		set_status_installed
-		sleep infinity
```
why dropping the `sleep infinity` ?

ANJANA-ARK replied:
```
install

# Install addon artifacts, if ADDON_IMAGE is present
[ -n "${ADDON_IMAGE:-}" ] && /scripts/osc-kata-addons-install.sh install

sleep infinity
```

here I've added sleep infinity after the execution of osc-kata-addons-install.sh script, if  not when install is called it goes to sleep infinity and rest of the script is not getting executed

gkurz replied:
This is typically a change that should be made in a preparatory patch :
1. a commit that moves `sleep infinity` to the caller at the right place, without changing behavior
2. current commit then only needs to focus on calling the new script

ANJANA-ARK replied:
Sure I've changed it to 2 commits

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -109,42 +109,41 @@ install() {
 	if [[ ${#install_rpms[@]} -eq 0 ]]; then
-		set_status_installed
-		sleep infinity
+	    	set_status_installed
```
Broken indentation.

ANJANA-ARK replied:
sure

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -109,42 +109,41 @@ install() {
+   	    	rm -rf /host/tmp/extensions/
+
+    		# Wait again: rpm-ostree install stages changes, requiring a reboot
+   		wait_for_reboot_clear
```
Same here, if you really need to move the lines in the `else` branch, please fix indentation.

ANJANA-ARK replied:
sure

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -166,7 +165,23 @@ uninstall() {
+  | sed 's/.*\[\(.*\)\].*/\1/' \
+  | tr ',' '\n' \
+  | sed 's/"//g; s/^[ \t]*//; s/[ \t]*$//;/^$/d'
+)
```
A comment to explain the intent behind this `sed` black magic would be greatly appreciated.

ANJANA-ARK replied:
sure, I'll add a comment. This is to get the list of "layered" packages (requested-local-packages) from rpm-ostree. https://github.com/openshift/sandboxed-containers-operator/pull/1334#discussion_r2563650335

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -166,7 +165,23 @@ uninstall() {
+   	 	else
+        		echo "Skipping $pkg (not layered)"
+   		 fi
+	done
```
What's the justification for this change ? Please add comments here and in the commit changelog.

ANJANA-ARK replied:
sure I'll add in commit changelog, 
in the script we have `PACKAGES="capstone daxctl-libs edk2-ovmf ipxe-roms-qemu kata-containers libfdt libpmem libpng librdmacm ndctl-libs pixman qemu-img qemu-kvm-common qemu-kvm-core seabios-bin seavgabios-bin virtiofsd"` 
here  few are not applicable for s390x, when the package is not present uninstall fails stating the package is not present. So using the layered packages sed command we're taking the packages which are present core OS and if any package which is given in `PACKAGES ` is not present here we're just skipping those, an alternative to sed command is using jq but we need to install jq on the RHCOS since we didn't want to install a new package from the script I've used sed command.

gkurz replied:
Please make the content of `PACKAGES` arch dependent instead of this logic that ignores uninstall errors.

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -211,18 +226,25 @@ main() {
+		# Install addon artifacts, if ADDON_IMAGE is present
+		[ -n "${ADDON_IMAGE:-}" ] && /scripts/osc-kata-addons-install.sh install
+
+		sleep infinity
```
Is this the `sleep infinity` that got removed in the other change ?

ANJANA-ARK replied:
yes

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -211,18 +226,25 @@ main() {
 		;;
 	*)
-		echo "Usage: $0 {install|uninstall}"
+        echo "Usage: $0 {install|uninstall}"
```
This change doesn't seem to be relevant for this PR.

beraldoleal reviewed (commented):
While we don't have an explicit rule that we should have >1 approval, its a good practice to solve all reviewer comments (or find an agreement) before merging.

I'm removing my LGTM, just to avoid this getting merged by accident... We are aware of the importance of this PR but I still see a few open comments from @gkurz that needs to be addressed before this gets in. @ANJANA-ARK  could you please fix them so we can merge this ASAP?

gkurz commented on scripts/kata-install/osc-kata-addons-install.sh:
```diff
@@ -0,0 +1,140 @@
+#   0 on success
+#######################################
+install_addons() {
+    local addon_image="${ADDON_IMAGE:-}"
```
Since this is mandatory, make it `${ADDON_IMAGE:?}`. The script will print `ADDON_IMAGE: parameter null or not set` and exit 1.

gkurz commented on scripts/kata-install/osc-kata-addons-install.sh:
```diff
@@ -0,0 +1,140 @@
+    local kernel_path=""
+    
+    # Extract and install kernel
+    if [ -n "$kernel_src" ]; then
```
With the check above, the `if` isn't needed anymore.

gkurz commented on scripts/kata-install/osc-kata-addons-install.sh:
```diff
@@ -0,0 +1,140 @@
+    local install_dir="/etc/kata-containers"
+    echo "Uninstalling addon artifacts"
+    
+    local kernel_src="${ADDON_KERNEL_PATH:-}"
```
AFAICT this is mandatory as well, `"${ADDON_KERNEL_PATH:?}"` and you can drop `[ -n "$kernel_src" ]` below, that becomes redundant.

beraldoleal commented on scripts/kata-install/Dockerfile:
```diff
@@ -6,10 +6,9 @@ ADD 50-kata-remote configuration-remote.toml /files/
-ADD osc-kata-install.sh osc-configs-script.sh osc-log-level.sh lib.sh /scripts/
+ADD osc-kata-install.sh osc-configs-script.sh osc-log-level.sh lib.sh osc-kata-addons-install.sh /scripts/
 
 RUN curl -sSL "https://github.com/opencontainers/umoci/releases/download/v0.4.7/umoci.amd64" -o "/usr/local/bin/umoci" &&\
```
@ANJANA-ARK please, help me to understand something...

This is something your PR is not touching, but the amd64 got my attention here... 

umoci v0.4.7 only has amd64 binaries. Is this intended to work only on x86_64 clusters? Or is this work focused on s390x as well? Will this work?

ANJANA-ARK replied:
Hi Beraldo,
Thanks for bringing this up..
We had a discussion on this when we started this development item and the suggestion was to enable Konflux for this daemonset image builds for GA and in the interim to share a s390x image built based on the PR once it is merged. 
For dev-test we have manually built a diff image and tested it, once the PR is merged, will share the container image built on s390x.
https://issues.redhat.com/browse/KATA-4307
Thanks..

beraldoleal replied:
Ok, just to make sure you are aware. Maybe you could consider adapting this Docker file to handle both archs.

gkurz replied:
This is definitely needed for the release, in another PR of course.

ANJANA-ARK replied:
Sure

ANJANA-A-R-K committed:
daemonset: Move `sleep infinity` from install() to main()

Sleep was inside install(), blocking any post-install steps. Moving it to
main() lets callers run additional scripts before the DaemonSet pauses.

Signed-off-by: ANJANA-A-R-K <anjana.a.r.k1@ibm.com>

gkurz commented on scripts/kata-install/osc-kata-install.sh:
```diff
@@ -164,6 +164,12 @@ uninstall() {
+	if [ "$ARCH" = "s390x" ]; then
+		PACKAGES="capstone kata-containers libfdt libpng pixman qemu-img qemu-kvm-common qemu-kvm-core virtiofsd"
+	fi
+
```
The package list for s390x should be defined at the same place as the one for x86_64, with a  `case...esac`, so that multi-arch support is explicit and unmissable. Non supported archs should cause the script to fail. Typically a change to do in a preliminary patch :wink:

gkurz replied:
Rest LGTM. I won't ask for more changes when that's done.

ANJANA-ARK replied:
Sure @gkurz I've made these changes 👍

ANJANA-A-R-K committed:
daemonset: make package selection arch explicit via case statement

Add package list for s390x and fail on unsupported architectures.

Signed-off-by: ANJANA-A-R-K <anjana.a.r.k1@ibm.com>

ANJANA-A-R-K committed:
controllers: support addon artifact images

Add support for installing custom kernel artifacts from container images in DaemonSet mode. Introduces osc-kata-addons-install.sh script and ConfigMap-based configuration (kata-addon-artifacts).

Reuses extract_container_image from lib.sh.

Fixes: https://issues.redhat.com/browse/KATA-4234

Signed-off-by: ANJANA-A-R-K <anjana.a.r.k1@ibm.com>

gkurz reviewed (approved):
/lgtm

Thanks @ANJANA-ARK !

gkurz commented:
/ok-to-test
